internal/config/config.go

```diff
@@ -130,3 +104,179 @@ func RootGitRepoDir() (string, error) {
 func ProjectConfigFile() string {
 	return projectConfigName
 }
+
+func readAccessToken() (string, error) {
+	ring, err := openKeyring()
+
+	if errors.Is(err, keyring.ErrNoAvailImpl) {
+		accessToken, tokenErr := readAccessTokenPath()
+		return string(accessToken), tokenErr
+	}
+
+	item, err := ring.Get(keyringKey)
+	if err == nil {
+		return string(item.Data), nil
+	}
+
+	if errors.Is(err, keyring.ErrKeyNotFound) {
+		// Migrate to keychain
+		accessToken, tokenErr := readAccessTokenPath()
+		if len(accessToken) > 0 && tokenErr == nil {
+			return migrateAccessToken(ring, accessToken)
+		}
+		return "", nil
+	}
+
+	return "", err
+}
+
+func migrateAccessToken(ring keyring.Keyring, accessToken []byte) (string, error) {
+	err := ring.Set(keyring.Item{
+		Key:  keyringKey,
+		Data: accessToken,
+	})
+	if err != nil {
+		return "", err
+	}
+	path, err := accessTokenPath()
+	if err != nil {
+		return "", err
+	}
+	err = os.Remove(path)
```

Comment:
I'm not sure how I feel about automatically removing a file on behalf of users. Should the migration be opt-in? Do we consider this file to be an implementation detail of the CLI that users should never interact with?

Reply:
> Do we consider this file to be an implementation detail of the CLI that users should never interact with?

Yes, the file is an implementation detail and I think we should do this for the user so they immediately benefit from a better security mechanism being used.

Reply:
I'll defer to @fatih and @iheanyi too since I think they've both got a lot more time in this than I do.

Reply:
I'm worried this might break CI users. Sometimes automations happen by mounting a secret file or something alike, so if the path is in such a path and the CI expects it somehow, I'm not sure how it would behave. Another edge case, sometimes the disk for certain files are read-only, so this might fail as well? 

I wonder if we should add a prompt or a flag that removes it from their disk. Something like: "Would you like to remove your old token?" or a `--force-remove-auth` flag.

Reply:
Another feedback: I think we should not error here if we can't remove it. Because we already successfully obtained the secret. I think printing a warning such as `"Couldn't remove old access token due to error: ..."`

Reply:
@fatih I don't see how a CI system would have something that here is picked up as a working keyring? It would only do that if there's a working keyring implementation which normally is not the case.

Even though if there is a keyring, in general CI systems would treat disks as ephemeral? Like it would drop the secret into place, but if there's a working keyring it could move it there but it still works.

Then on a next CI run, the file would be put back in place again and things keep working?

The reason I'm pretty strong on removing it, is because without it the vast majority of cases would never remove the file then. It would be lingering there forever. 

> I think printing a warning such as `"Couldn't remove old access token due to error: ..."`

I think similarly here, I worry that people ignore a warning and that we still have a valid token on disk lingering around that still works.

Reply:
> The reason I'm pretty strong on removing it, is because without it the vast majority of cases would never remove the file then. It would be lingering there forever.

That's a good point. 

> Even though if there is a keyring, in general CI systems would treat disks as ephemeral? Like it would drop the secret into place, but if there's a working keyring it could move it there but it still works.

They would treat as ephemeral, thought some CI systems also just attach/mount an EBS for the underlying disk. It can go crazy. For GitHub actions or similar CI's you're right, but people with a more traditional setup, such as using Jenkins might not have the same environment.

> I think similarly here, I worry that people ignore a warning and that we still have a valid token on disk lingering around that still works.

That's a good call, but I think it might still break people's workflows. What if we do this: Let's not return an error, print a warning. Deploy `pscale` and remove the warning in like 1/2 weeks later. That would show us if we missed any edge cases.

Reply:
> They would treat as ephemeral, thought some CI systems also just attach/mount an EBS for the underlying disk. It can go crazy. For GitHub actions or similar CI's you're right, but people with a more traditional setup, such as using Jenkins might not have the same environment.

Yeah, but the same reasoning here applies? If the machine is a long term stateful thing and it has a keyring, it means that the keyring also would keep working?

> That's a good call, but I think it might still break people's workflows. What if we do this: Let's not return an error, print a warning. Deploy pscale and remove the warning in like 1/2 weeks later. That would show us if we missed any edge cases.

What about a more general changed approach here that does include removing the file (and erroring it fails), but we don't do that right now.

We first release the CLI now without immediately deleting it, to see if that causes any adverse effects for people. We then release a future new version in a few weeks that deletes the token from disk if it successfully reads from the keyring? 

I think that helps address all situations then?

Reply:
> Yeah, but the same reasoning here applies? If the machine is a long term stateful thing and it has a keyring, it means that the keyring also would keep working?

True, but there are also systems like in K8S where you can attach/mount stateful disks, even though the underlying compute instance might be not stateful (i.e: Pod). I don't know how various Keyring environments work of course, but I think being careful here wouldn't hurt us. 

> We first release the CLI now without immediately deleting it, to see if that causes any adverse effects for people. We then release a future new version in a few weeks that deletes the token from disk if it successfully reads from the keyring?
> I think that helps address all situations then?

I like it a lot, let's do it 👍🏼